src/odb/src/lefout/lefout.cpp

```diff
@@ -135,32 +138,81 @@ void lefout::writeHeader(dbBlock* db_block)
+void lefout::writeObstructions(dbBlock* db_block, int bloat_factor)
 {
-  std::set<dbTechLayer*> obstruction_layers;
-  getTechLayerObstructions(db_block, obstruction_layers);
+  std::map<dbTechLayer*,std::set<Rect*>> obstructions;
+  getObstructions(db_block, obstructions);
+  mergeObstructions(bloat_factor, obstructions);
 
   fprintf(_out, "  OBS\n");
-  dbBox* block_bounding_box = db_block->getBBox();
-  for (dbTechLayer* tech_layer : obstruction_layers) {
+
+  for (const auto& [tech_layer, rects_per_layer] : obstructions ) {
     fprintf(_out, "    LAYER %s ;\n", tech_layer->getName().c_str());
-    writeBox("  ", block_bounding_box);
+    for (Rect* rect : rects_per_layer) {
+      writeBox("  ", rect);
+    }
   }
+  
   fprintf(_out, "  END\n");
 }
 
-void lefout::getTechLayerObstructions(
+void lefout::mergeObstructions(
+  int bloat_factor,
+  std::map<dbTechLayer*,std::set<Rect*>>& obstructions) const
+{
+  using namespace boost::polygon::operators;
+
+  for (const auto& [tech_layer, rects_per_layer] : obstructions ) {
+    int bloat = bloat_factor*tech_layer->getPitch();
+
+    // Compose all the rects in the layer to merged polygon with bloat
+    boost::polygon::polygon_90_set_data<int> polySet;
+    for (Rect* curr_rect : rects_per_layer) {
+      boost::polygon::polygon_90_set_data<int> poly;
+      poly = boost::polygon::rectangle_data<int>{curr_rect->xMin(),
+                                                 curr_rect->yMin(),
+                                                 curr_rect->xMax(),
+                                                 curr_rect->yMax()};
+      polySet += poly.bloat(bloat, bloat, bloat, bloat);
+    }
+    polySet.shrink2(bloat, bloat, bloat, bloat);
+  
+    // Decompose the polygon set to rectanges in non-preferred direction
+    std::vector<boost::polygon::rectangle_data<int>> rects;
+    if (tech_layer->getDirection() == odb::dbTechLayerDir::HORIZONTAL) {
+      polySet.get_rectangles(rects, boost::polygon::VERTICAL); 
+    } else if (tech_layer->getDirection() == odb::dbTechLayerDir::VERTICAL) {
+      polySet.get_rectangles(rects, boost::polygon::HORIZONTAL); 
+    } else if (tech_layer->getDirection() == odb::dbTechLayerDir::NONE) {
+      polySet.get_rectangles(rects); 
+    }
+
+    if (rects.empty()) {continue;}
+
+    // Add the new rects to the obstructions map
+    obstructions.erase(tech_layer); 
+    for (const auto& new_rect : rects) {
+      Rect* obs_rect = new Rect(boost::polygon::xl(new_rect),
+                                boost::polygon::yl(new_rect),
+                                boost::polygon::xh(new_rect),
+                                boost::polygon::yh(new_rect));
+      obstructions[tech_layer].insert(obs_rect);
+    }
+  }
+}
+
+void lefout::getObstructions(
     dbBlock* db_block,
-    std::set<dbTechLayer*>& obstruction_layers) const
+    std::map<dbTechLayer*, std::set<Rect*>>& obstructions) const
 {
   for (dbNet* net : db_block->getNets()) {
```

Comment:
There is a pre-existing issue here that becomes more important with this change.  You are not picking up and shapes from instances that acts as obs as well.  You can see it by looking at m1 which only shows the power straps.

You should add a loop over db_block->getInsts() and add the pins and obs shapes to obstructions

Reply:
I should block the entire inst or only his obs with inst->getBBox()->getObstructions?

Reply:
The obstructions of the insts is the nets/wires of the insts? same as the nets of the block?

Reply:
The instances are in addition to the nets.  pins & obs of instances also block routing just like nets do.

Reply:
For example in this area these are the shapes belonging to nets:
![image](https://user-images.githubusercontent.com/761514/190003726-09dc1b29-d10c-4689-95af-dd93403bc66c.png)
but there are also these shapes belonging to the instance:
![image](https://user-images.githubusercontent.com/761514/190003782-e3a9d00a-b435-4aa8-a47c-1e75c38c9700.png)